### ISSUE TITLE
Start scripts not working for certain devices

### ISSUE THREAD

**Issue by dnikles** [maykar/plex_assistant#93]
I installed the latest version. I am using IFTTT. Here is a relevant portion of the log:
```
2021-02-12 16:54:28 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] IFTTT Call: play Christmas vacation on living room
2021-02-12 16:54:28 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Command: play Christmas vacation on living room
2021-02-12 16:54:29 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Processed Command: media = christmas vacation, device = living room
2021-02-12 16:54:49 WARNING (SyncWorker_1) [custom_components.plex_assistant.const] Cast device not found: "Living Room".
```

If I set the default device to one of the entities that the plex integration creates, for example media_player.plex_plex_for_roku_living_room and call plex assistant it will play on the device if plex is already open. I renamed the friendly name of this device to  "Living Room" but still get the above error. However, I do have mutiple friendly names of "living room" including for example media_player.living_room which is what I use launch the plex app. I wonder if some sort of mapping like you have where we map a script to a device would work to map a streaming device to a friendly device name.

**Comment by maykar** [comment]
So, I don't think the multiple names would be an issue. Roku devices wouldn't be seen as a player for Plex Assistant as it only looks for plex and google devices.

If you wouldn't mind I have some additional debug that you can add so we can see what devices are listed.

On line 162 of `__init__.py` add `_LOGGER.warning(pa.devices)` before that line, so it will look something like this:

```
        _LOGGER.warning(pa.devices)

        if device[1] < 60 or not responding:
```

This will post a warning in your logs of every device and their info. I'll try to add something like this as a debug log in the future, but for now this will give us at least an idea of what devices are being seen.

**Comment by dnikles** [comment]
so my thought about needing to map to the entity plex provides is no good. Here's a relevant log portion. I first tried naming the device. Second time I did not name the device, but the default device did have plex open. Neither worked because HA and/or plex didn't see the player yet I suppose. I did get it to work once in my kitchen, but as you can see it sees that entity now.

```                     '"plex_assistant.command", "command": "play Big Lebowski '
2021-02-13 12:34:03 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event ifttt_webhook_received[L]: action=call_service, service=plex_assistant.command, command=play Big Lebowski on living room, webhook_id=4fc3d0e0ec50cb54d8486d302ceb7a934dc98e81700aba4517abffcbf3e36bea>
2021-02-13 12:34:03 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] IFTTT Call: play Big Lebowski on living room
2021-02-13 12:34:03 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=plex_assistant, service=command, service_data=command=play Big Lebowski on living room>
2021-02-13 12:34:03 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Command: play Big Lebowski on living room
2021-02-13 12:34:03 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Processed Command: media = big lebowski, device = living room
2021-02-13 12:34:24 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] PA Devices: {'Kitchen tv': {'entity_id': 'media_player.plex_plex_for_roku_kitchen_tv', 'device_type': 'plex'}, 'Living Room mini': {'entity_id': 'media_player.living_room_mini', 'device_type': 'cast'}, 'Dining room mini': {'entity_id': 'media_player.dining_room_mini', 'device_type': 'cast'}, 'Soundbar': {'entity_id': 'media_player.soundbar', 'device_type': 'cast'},  'Master Bedroom speaker': {'entity_id': 'media_player.master_bedroom_speaker', 'device_type': 'cast'},  'Basement speaker': {'entity_id': 'media_player.basement_speaker', 'device_type': 'cast'}, 'Kitchen speaker': {'entity_id': 'media_player.kitchen_speaker', 'device_type': 'cast'}}
2021-02-13 12:34:24 WARNING (SyncWorker_1) [custom_components.plex_assistant.const] Cast device not found: "Living Room".
                     '"plex_assistant.command", "command": "play Big Lebowski" '
2021-02-13 12:35:29 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event ifttt_webhook_received[L]: action=call_service, service=plex_assistant.command, command=play Big Lebowski, webhook_id=4fc3d0e0ec50cb54d8486d302ceb7a934dc98e81700aba4517abffcbf3e36bea>
2021-02-13 12:35:29 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] IFTTT Call: play Big Lebowski
2021-02-13 12:35:29 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=plex_assistant, service=command, service_data=command=play Big Lebowski>
2021-02-13 12:35:29 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] Command: play Big Lebowski
2021-02-13 12:35:29 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] Processed Command: media = big lebowski, device = Plex (Plex for Roku - Living room)
2021-02-13 12:35:50 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] PA Devices: {'Kitchen tv': {'entity_id': 'media_player.plex_plex_for_roku_kitchen_tv', 'device_type': 'plex'}, 'Living Room mini': {'entity_id': 'media_player.living_room_mini', 'device_type': 'cast'}, 'Dining room mini': {'entity_id': 'media_player.dining_room_mini', 'device_type': 'cast'}, 'Soundbar': {'entity_id': 'media_player.soundbar', 'device_type': 'cast'},  'Master Bedroom speaker': {'entity_id': 'media_player.master_bedroom_speaker', 'device_type': 'cast'}, 'Basement speaker': {'entity_id': 'media_player.basement_speaker', 'device_type': 'cast'}, 'Kitchen speaker': {'entity_id': 'media_player.kitchen_speaker', 'device_type': 'cast'}} ```

**Comment by dnikles** [comment]
Here's a few more observations. I thought maybe I could get around this by having my "start script" wait until the device is available before proceeding, but it seems the script is called but does not wait for the script to finish. The main issue still seems to be getting plex, and then home assistant, to see that the client is available. I was trying to trigger this previously by sending command: update sensor to plex_assistant.command.  

I also believe I do have to have the device seen by the roku and then manually change the friendly name, otherwise it shows up as Plex (Plex for Roku Kitchen tv). That's an easy fix on the end user side. 

Lastly, if the device does show up as an available plex entity, it seems the start script is not called. So if you are in plex, then exit it on the roku, it will not launch plex for you, but as soon as you launch it , it will try to play the movie you requested.

**Comment by maykar** [comment]
So to give a bit more insight into how the start script feature works:

* When you make a call with plex assistant it makes a list of all available and compatible (plex or google cast) devices
* Then if the device name is in the start script config it will check if it's one of the available devices and responding to commands
* If either of those checks fail it then runs the start script and continues with the following
* Every 2 seconds for up to ~40 seconds it updates the available device list and checks if the selected device is responding
* If the device is responding it will stop the previous check and then give it a few seconds for startup before continuing
* Otherwise it keeps trying until those ~40 seconds are up

Then, if the device is not responding and not in the list of available devices it stops and gives the device not found error 

https://github.com/maykar/plex_assistant/blob/4249ba041d7e0ec1e855a9b56ff5f8bf65b9955d/custom_components/plex_assistant/__init__.py#L145

So there is no need to have your script wait, it just needs to be able to start the Plex app.

I, unfortunately, don't have a Roku device to test this with.

After making the call and the start script runs, if you go to your developer tools and look at the state of the device is it still listed as "unavailable"?

**Comment by dnikles** [comment]
If I go to developer tools while a device is unavailable, it isn't even available for me to pick for it to show "unavailable". Even if I paste the name in it doesn't come up. However, if I get to it from the plex integration, it does show This entity is not currently available. I waited several minutes while in plex on a roku client and the entity did not become available. However, the moment I reloaded the plex integration, the client showed as idle immediately.

**Comment by dnikles** [comment]
Sounds like you are already doing something like this https://www.home-assistant.io/integrations/plex/#service-plexscan_for_clients so maybe there is a bug in plex.scan_for_clients in the plex integration. I haven't tested thoroughly yet though. I saw the client show up as soon as I reloaded the plex integration, but I did not try manually calling plex.scan_for_clients to see if that would make it appear. And once it does show up I don't know how to make it unavailable to test again easily.

**Comment by dnikles** [comment]
testing on another roku, and calling the plex.scan_for_clients service had it show up immediately.

**Comment by maykar** [comment]
Cool, so I was doing something similar, but maybe it's not as effective as the service call. Here's another change you can try in `__init__.py`:

Replace these lines:
```
        if device[0] in start_script_keys:
            if device[0] not in pa.device_names or not responding:
                timeout = 0
                hass.services.call("script", start_script[device[0]].replace("script.", ""))

                while timeout < 20 and not responding:
                    if (timeout % 2) == 0:
                        ha_plex_server._async_update_platforms()
```

with this and restart:
```
            if device[0] not in pa.device_names or not responding:
                timeout = 0
                hass.services.call("script", start_script[device[0]].replace("script.", ""))
                hass.services.call("plex", "scan_for_clients")

                while timeout < 20 and not responding:
                    if (timeout % 5) == 0:
                        hass.services.call("plex", "scan_for_clients")
```

**Comment by maykar** [comment]
This still works with the clients that I've tested, so if this works for your Roku's then I think we have a winning solution.

**Comment by dnikles** [comment]
So close! It looks like the device was found almost immediately but it still said cast device not found. Maybe it needs an additional second to sleep before trying cast? When I said the command again it went through.

**Comment by maykar** [comment]
Try increasing that sleep like we did at the start of all this and maybe add a `get_devices(hass, pa)` and a debug in there to make sure that's getting hit.

```
                if responding:
                    _LOGGER.debug("Device Responding")
                    get_devices(hass, pa)
                    time.sleep(3)
```

**Comment by dnikles** [comment]
```
021-02-14 09:36:30 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] Command: play cars on living room Roku
2021-02-14 09:36:30 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] Processed Command: media = cars, device = living room roku
2021-02-14 09:36:51 DEBUG (SyncWorker_4) [custom_components.plex_assistant.const] PA Devices: {'Living Room mini': {'entity_id': 'media_player.living_room_mini', 'device_type': 'cast'}, 'Master Bedroom speaker': {'entity_id': 'media_player.master_bedroom_speaker', 'device_type': 'cast'}, 'Basement speaker': {'entity_id': 'media_player.basement_speaker', 'device_type': 'cast'}, "'s room speaker": {'entity_id': 'media_player._s_room_speaker', 'device_type': 'cast'}, "'s room speaker": {'entity_id': 'media_player._s_room_speaker', 'device_type': 'cast'}, "'s room speaker": {'entity_id': 'media_player._s_room_speaker', 'device_type': 'cast'}, 'Kitchen speaker': {'entity_id': 'media_player.kitchen_speaker', 'device_type': 'cast'}, 'Soundbar': {'entity_id': 'media_player.soundbar', 'device_type': 'cast'}, 'Dining room mini': {'entity_id': 'media_player.dining_room_mini', 'device_type': 'cast'}, 'Living Room roku': {'entity_id': 'media_player.plex_plex_for_roku_living_room', 'device_type': 'plex'}}
2021-02-14 09:36:51 WARNING (SyncWorker_4) [custom_components.plex_assistant.const] Cast device not found: "Living Room Roku".
```

Couple of interesting things here. I have time.sleep(30), but it's only 21 seconds from  processed command to Cast device not found. And the cast device that is not found is listed on the line before. It's like it needs a moment to wait after it sees the entity available. I also don't see "Device Responding" anywhere in the logs. I think I put it in the code in the right spot, and I restarted home assistant after saving the change.
```
                while timeout < 20 and not responding:
                    if (timeout % 5) == 0:
                        hass.services.call("plex", "scan_for_clients")
                    responding = device_responding(hass, pa, device[0])
                    timeout += 1
                    time.sleep(1)

                if responding:
                    _LOGGER.debug("Device Responding")
                    get_devices(hass, pa)
                    time.sleep(30)

                device = fuzzy(command["device"] or default_device, list(pa.devices.keys()))

        _LOGGER.debug("PA Devices: %s", pa.devices)
```

**Comment by maykar** [comment]
Here's another shot, I'm thinking that the roku isn't responding to the "responding" check, but does in fact appear:

```
                while timeout < 30 and not responding and device[0] not in pa.device_names:
                    if (timeout % 10) == 0):
                        hass.services.call("plex", "scan_for_clients")
                    responding = device_responding(hass, pa, device[0])
                    time.sleep(1)
                    timeout += 1

                responding = responding or device[0] in pa.device_names

                if responding:
                    _LOGGER.debug("Device Responding")
                    time.sleep(3)

                device = fuzzy(command["device"] or default_device, list(pa.devices.keys()))

        _LOGGER.debug("PA Devices: %s", pa.devices)
```

**Comment by maykar** [comment]
This kind of defeats the purpose of the responding check, but I unfortunately don't have a roku to test this on

**Comment by maykar** [comment]
I might just need to rethink the responding function

**Comment by dnikles** [comment]
> This kind of defeats the purpose of the responding check, but I unfortunately don't have a roku to test this on. The responding check was put in place since a device can be listed in plex as a valid device (so it would be in pa.device_names), but actually be unavailable. The Plex API can take up to 15 minutes to mark a device as unavailable after it has been turned off or closed.
> 
> You can see it in developer tools, if you close your Plex app on the roku it will still list it as "idle" when in fact it should be "unavailable"

Yeah, this is why I'd still want the start script to run even if the device is "available".  If in those 15 minutes you run the scan_for_clients, I wonder if it would show the client unavailable?  I'll test that out.

**Comment by maykar** [comment]
I just retooled the responding function and put it in a dev branch. If you'd like to give it a try you can just copy/replace all of `__init__.py` with [this](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/__init__.py) and do the same with `helpers.py` with [this](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/helpers.py)

**Comment by maykar** [comment]
> Yeah, this is why I'd still want the start script to run even if the device is "available". If in those 15 minutes you run the scan_for_clients, I wonder if it would show the client unavailable? I'll test that out.

That's the point of the responding function, it's what checks to be sure the clients are actually available. In my experience the plex.scan_for_clients service doesn't solve the issue. It still shows unavailable clients as idle.

**Comment by dnikles** [comment]
I tried the branch but I get errors now. It isn't calling the start script for me now either. 
```
2021-02-14 12:23:20 DEBUG (SyncWorker_8) [custom_components.plex_assistant.const] IFTTT Call: play cars on kitchen TV
2021-02-14 12:23:20 DEBUG (SyncWorker_5) [custom_components.plex_assistant.const] Command: play cars on kitchen TV
2021-02-14 12:23:20 DEBUG (SyncWorker_5) [custom_components.plex_assistant.const] Processed Command: media = cars, device = kitchen tv
  File "/config/custom_components/plex_assistant/__init__.py", line 148, in handle_input
  File "/config/custom_components/plex_assistant/helpers.py", line 47, in device_responding
```

**Comment by maykar** [comment]
sorry, but what's the full error there. it's telling me where they are but not what they are

**Comment by dnikles** [comment]
Hey sorry I was just grepping home-assistant.log for custom_components.plex_assistant . Maybe this is what you're after:
```
2021-02-14 12:10:04 ERROR (MainThread) [homeassistant.core] Error executing service: <ServiceCall plex_assistant.command (c:be2ade49b55b9fc41cea22e97d75ac41): co>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/core.py", line 1471, in catch_exceptions
    await coro_or_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1494, in _execute_service
    await self._hass.async_add_executor_job(handler.job.target, service_call)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/plex_assistant/__init__.py", line 148, in handle_input
    responding = device_responding(hass, pa, device[0]) if device[0] in start_script_keys else True
  File "/config/custom_components/plex_assistant/helpers.py", line 47, in device_responding
    device = hass.data["media_player"].get_entity(pa.devices[device]["entity_id"]).device
KeyError: 'Kitchen TV'
```

**Comment by maykar** [comment]
Updated with a fix and added some debugging lines just in case: [`__init__.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/__init__.py) [`helpers.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/helpers.py)

**Comment by dnikles** [comment]
This time I watched the entity in the device in home assistant, and it became "idle" just a second or two after the plex channel finished loading on the roku. And then there was a delay before we tried playing the movie, but it still showed cast device not found. I tried again and 30 second after the cast device not found, the movie was playing for me. Some of that delay would be in me deciding to say the command again, and a little more delay for the command to get from google to ifttt and back to home assistant.  Would it be possible to wait 5 seconds if cast device not found, then retry, and loop that for 5 attempts or so? I think it's a timing issue either with plex or with the roku client, and I'm just missing it.
```
2021-02-14 13:40:50 DEBUG (SyncWorker_6) [custom_components.plex_assistant.const] IFTTT Call: play cars on kitchen TV
2021-02-14 13:40:50 DEBUG (SyncWorker_8) [custom_components.plex_assistant.const] Command: play cars on kitchen TV
2021-02-14 13:40:50 DEBUG (SyncWorker_8) [custom_components.plex_assistant.const] Processed Command: media = cars, device = kitchen tv
2021-02-14 13:41:31 DEBUG (SyncWorker_8) [custom_components.plex_assistant.const] PA Devices: { 'Kitchen tv': {'entity_id': 'media_player.plex_plex_for_roku_kitchen_tv', 'device_type': 'plex'}}
2021-02-14 13:41:31 WARNING (SyncWorker_8) [custom_components.plex_assistant.const] Cast device not found: "Kitchen Tv".
2021-02-14 13:42:02 DEBUG (SyncWorker_9) [custom_components.plex_assistant.const] IFTTT Call: play cars on kitchen TV
2021-02-14 13:42:02 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Command: play cars on kitchen TV
2021-02-14 13:42:02 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Processed Command: media = cars, device = kitchen tv
2021-02-14 13:42:02 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] PA Devices: { 'Kitchen tv': {'entity_id': 'media_player.plex_plex_for_roku_kitchen_tv', 'device_type': 'plex'}}
2021-02-14 13:42:02 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Device: Kitchen tv
2021-02-14 13:42:03 DEBUG (SyncWorker_1) [custom_components.plex_assistant.const] Media: <Movie:4479:Cars>
```

**Comment by maykar** [comment]
So, it looks like Roku's just won't respond in any way that I can use to tell if it's actually idle or if it's unavailable. I'll have to think on solutions as the last thing I want is for anything that uses a start script to always have a long delay after each command, whether it's available or not.

Maybe I can try to catch the service error on control commands and play commands and work with that.

**Comment by dnikles** [comment]
> Maybe I can try to catch the service error on control commands and play commands and work with that.

I think that's what I was suggesting, but maybe I'm off. Try, try again is what I think might work. Rather than keep waiting, keep trying until it works. Instead of testing if the client is available, test if the play command succeeds. Finally give up when there's no reasonable hope it will work.

**Comment by dnikles** [comment]
RE: the topic change, the start script is working for me after replacing __init__.py helpers.py with the latest versions you suggested. What I'm seeing is once the device shows up in the PA Device list, it will work soon, but won't necessarily work yet.

**Comment by maykar** [comment]
>  it will work soon, but won't necessarily work yet

Could you expand on that a bit, I'm not exactly sure what you're saying.

**Comment by dnikles** [comment]
Sure! So when I call plex assistant through IFTTT, the entity provided by the plex integration in HA shows as unavailable. Plex assistant calls the scan_for_clients and almost immediately the entity will show as idle. We'd like to think that means the device is ready to play a video if we tell plex to do so, but we've seen that's not the case. However, from what I've seen, when the entity goes from unavailable to idle, we will _soon_ be able to have plex play a video on it. We just don't know how soon that is, 10 seconds? 40? This is different than me requesting it to for example play on garage tv, when a garage tv doesn't exist at all. Garage tv will never be found by scan_for_clients. So, once scan_for_clients sees that for example kitchen tv is a PA Device, we can be reasonably sure we will be able to play a video on it, we just might not be able to do it immediately. So if we try to play the video and it fails to do so (Cast device not found), I think we should keep trying until successful. Looking at my last log, I told it to play a movie, it failed. I told it again, and it worked. The second time I told it, it didn't run the start script because it already saw the device as available, so it just tried to play the movie again and was successful.

**Comment by maykar** [comment]
Ah okay cool, but this is all with the dev branch that essentially removes "responding".

I didn't add that to the latest release as it's not really a "fix" yet. Will continue to experiment and will let you know when I have something to check out.

**Comment by dnikles** [comment]
I think what I said is the case in both dev and master, because I'm pretty certain it would apply to my comment [here](https://github.com/maykar/plex_assistant/issues/93#issuecomment-778789245) which was before I tried the dev branch. We see in the log it saw the device in PA Devices but returned Cast device not found.

**Comment by maykar** [comment]
The problem with doing something like trying attempting to play on the device until it succeeds is that each time it tries, it causes and error in the log on fail. This is an error I can't catch and I don't want to fill logs with errors just for this feature, so I'm trying to find a method that will fail on unresponsive devices with an error that I can catch and act on.

**Comment by maykar** [comment]
What types of Roku devices are you using? I may just have to get one, I've thought about grabbing one for the wife in the past. Maybe now's the time.

**Comment by dnikles** [comment]
Oh shoot I thought "Cast device not found" was coming from plex assistant! I have a few tcl 4k roku tv's, a roku streaming stick (no longer made) and an even older roku 3. I bet I can make this work for my purposes by putting a sleep right after when we list the PA Devices, but we wouldn't want to introduce that delay to users of other devices.

**Comment by maykar** [comment]
"Cast device not found" is coming from Plex Assistant, but when you see that error Plex Assistant has stopped so the HA media commands are never sent. I have another idea I'm testing out, will let you know when it's ready.

**Comment by maykar** [comment]
You having the same issues on all of your roku devices?

**Comment by dnikles** [comment]
I haven't tested on the 3, but yes all of the others have the same issue.

**Comment by dnikles** [comment]
I don't know if this is helpful but I wanted to test it out. I made a script purely in home assistant to open plex, check if the roku plex client is available, and once it is play a specific movie to it. I waited until the device was unavailable, called this script, and it worked with no real delay as well. So I'm wondering if just like scan_for_clients seems to work well, if we should try media_player.play_media ? Then again maybe I just got lucky this time.
```
alias: Start Plex Kitchen
sequence:
  - service: media_player.select_source
    data:
      source: Plex - Stream for Free
    entity_id: media_player.kitchen_tv
  - repeat:
      while:
        - condition: state
          entity_id: media_player.plex_plex_for_roku_kitchen_tv
          state: unavailable
      sequence:
        - service: plex.scan_for_clients
          data: {}
        - delay:
            hours: 0
            minutes: 0
            seconds: 1
            milliseconds: 0
  - service: media_player.play_media
    data:
      media_content_type: movie
      media_content_id: '{ "library_name": "Movies", "title": "Cars" }'
      entity_id: media_player.plex_plex_for_roku_kitchen_tv
    entity_id: media_player.plex_plex_for_roku_kitchen_tv
mode: single
```

**Comment by maykar** [comment]
The issue here isn't playing after opening the app, that parts easy enough. I could make that work right now. The issue is that there needs to be a reliable method to know if the device is actually available. Plex's API and (as a consequence of that) HA's media player entities stay listed as available long after they've been closed.

This means that after you start the app the first time, go do something else on the device or the Plex app goes idle and quits, then when you go to use Plex Assistant again there is no way to know if it should open the app again. It will assume it's open since both Plex and HA claim it is. Then it will not run the script and will error out on play.

Plex assistant uses the play_media service already, but calling that service on an unavailable device will cause an error in the log one that I cannot catch/prevent. Which is why I mentioned earlier that repeatedly calling it isn't an option, since it would fill the log with service errors.

**Comment by maykar** [comment]
The current solution works for all clients I can test, but not Roku's apparently.

Earlier today I got a Roku Express and will test when I can.

**Comment by dnikles** [comment]
Sorry I'm sure I'm misinterpreting how it works. Plex assistant only seems to get a movie to play for me on the second attempt, even if the entity shows available for a bit thanks to calling sleep to wait, whereas in the yaml I posted it plays immediately once the entity is available, which has only taken a second or two.

I've also seen scan for clients will move the entity to unavailable within a few minutes of the client disconnecting. So if you call that before checking for clients then the window for having closed a Plex client and not being able to identify that it needs to be opened again is pretty small.

**Comment by maykar** [comment]
Your devices don't respond to the "responding" check correctly. This is what is resulting in the strange behavior you see.

**Comment by dnikles** [comment]
Is the responding check more than checking that the device isn't unavailable? The home assistant script I have is working 100% for me, and the the movie starts playing within seconds of starting the script, almost as soon as plex loads. When I'm seeing faillures, I see the entity go from unavailable -> idle, then I can wait 20 or more seconds, and it still fails to play. That's why I thought it may be related to how we are calling the playback, but if we are both using media_player.play_media I don't have an explanation. What I'm testing is not related to the state still showing available after it's been closed. I'm always starting from unavailable.

**Comment by maykar** [comment]
> Is the responding check more than checking that the device isn't unavailable?

Yes, I can't rely on HA to tell me if the device is available or not. I've attempted a few things with it so far, but now that I have a Roku device I can more thoroughly test what might work. I've been trying to connect to the device directly using some methods from PlexAPI and others to get a failing response from a device that falsely reports as available in Plex and in HA.

**Comment by maykar** [comment]
Found a method that works on my Roku and seems pretty reliable, mind giving it a spin?

[`__init__.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/__init__.py) [`helpers.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/helpers.py)

**Comment by dnikles** [comment]
After a couple tests this has so far worked!

**Comment by maykar** [comment]
I noticed that, at least on my Roku, telling it to source Plex even when Plex was already opened seemed to cause the Plex app to quit and reopen. So, I improved the start script to check if it's already open first. Might be useful for you as well:

```
start_roku_plex:
  sequence:
    - condition: template
      value_template: "{{ state_attr('media_player.roku', 'source') != 'Plex - Stream for Free' }}"
    - service: media_player.select_source
      entity_id: media_player.roku
      data:
        source: Plex - Stream for Free
```

**Comment by maykar** [comment]
Think it might be safe to say this one is closed (just released an update with these changes). Feel free to continue the conversation here if you have any more questions or anything.

**Comment by dnikles** [comment]
> I noticed that, at least on my Roku, telling it to source to Plex when it was already open caused the Plex app to quit and reopen (screen flash to blue, then back to Plex). So, I improved the start script to check if it's already open first. Might be useful for you as well:
> 
> ```
> start_roku_plex:
>   sequence:
>     - condition: template
>       value_template: "{{ state_attr('media_player.roku', 'source') != 'Plex - Stream for Free' }}"
>     - service: media_player.select_source
>       entity_id: media_player.roku
>       data:
>         source: Plex - Stream for Free
> ```

True, more than one way to skin a cat too. This also turns on my tv first if it is off
```
alias: Start Plex Bedroom
sequence:
  - choose:
      - conditions:
          - condition: state
            entity_id: media_player.bedroom_roku
            state: standby
        sequence:
          - service: media_player.turn_on
            data: {}
            entity_id: media_player.bedroom_roku
    default: []
  - choose:
      - conditions:
          - condition: state
            entity_id: media_player.bedroom_roku
            state: Plex - Stream for Free
            attribute: source
        sequence: []
    default:
      - service: media_player.select_source
        data:
          source: Plex - Stream for Free
        entity_id: media_player.bedroom_roku
mode: single
```

**Comment by dnikles** [comment]
sorry to say but this still isn't reliable for me. Just tried on my living room tv. It did not list it in PA devices, then got cast device not found. Minutes later and the entity still shows unavailable. I'm sure if I scan for clients it will show up right away. 

```
2021-02-16 19:48:56 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] PA Devices: {'Kitchen speaker': {'entity_id': 'media_player.kitchen_speaker', 'device_type': 'cast'}, 'Soundbar': {'entity_id': 'media_player.soundbar', 'device_type': 'cast'}, 'Dining room mini': {'entity_id': 'media_player.dining_room_mini', 'device_type': 'cast'}, 'Living Room mini': {'entity_id': 'media_player.living_room_mini', 'device_type': 'cast'}, " 's room speaker": {'entity_id': 'media_player. _s_room_speaker', 'device_type': 'cast'}, 'Basement speaker': {'entity_id': 'media_player.basement_speaker', 'device_type': 'cast'}, " 's room speaker": {'entity_id': 'media_player. _s_room_speaker', 'device_type': 'cast'}, 'Master Bedroom speaker': {'entity_id': 'media_player.master_bedroom_speaker', 'device_type': 'cast'}, " 's room speaker": {'entity_id': 'media_player. _s_room_speaker', 'device_type': 'cast'}}
2021-02-16 19:48:56 WARNING (SyncWorker_2) [custom_components.plex_assistant.const] Cast device not found: "Living Room Tv".
```

**Comment by maykar** [comment]
Keep an eye on it and see if you can try to nail down why it wouldn't be seen. As far as scanning for clients:

if the device is unavailable the start script is called and then after 5 seconds we do [plex.scan_for_clients](https://github.com/maykar/plex_assistant/blob/4d3cc1e7dd41a5a356c71e12284f18c131d748b1/custom_components/plex_assistant/__init__.py#L158)

Then if it's not found after 30+ seconds we [call it again](https://github.com/maykar/plex_assistant/blob/4d3cc1e7dd41a5a356c71e12284f18c131d748b1/custom_components/plex_assistant/__init__.py#L166)

**Comment by maykar** [comment]
Here is another slightly tweaked version to try, its just the one file:  [`__init__.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/__init__.py)

Problem is that I can't keep doing tweak after tweak, I don't have your exact setup so I can't know what's going to help and what's not. I can't see what's going on or test what works on that specific device.

I even picked up a Roku and it's working great (not that I regret getting one, it'll get a ton of use and it was mostly funded by your generous donation. Thank you), but apparently there's enough of a difference between Roku models that one can work great and another fails.

TL;DR I'm not sure how many more small tweaks to the code I can do here. I have a total of 7 different devices with Plex clients that I'm testing with. Aside from very early attempts, every version of this has worked on all of those devices.

**Comment by maykar** [comment]
Something you can adjust to see if it helps are adjusting the 2 sleep timers. One on line 159 and one on 179. They should both be around the same amount. The first one is the delay before refreshing/scanning clients after the script is called if the device isn't listed. The second one is how much time to wait to continue before playing/controlling a device that was listed as available, but was actually unavailable (essentially how long it takes for the Plex app start responding after opening).

**Comment by maykar** [comment]
Just as an update, if you tried it already I've tweaked it even more. Can use the same link above for `__init__.py`

**Comment by dnikles** [comment]
> Problem is that I can't keep doing tweak after tweak, I don't have your exact setup so I can't know what's going to help and what's not. I can't see what's going on or test what works on that specific device.

Hey I totally understand, you don't owe me anything! I still don't understand why the purely yaml home assistant script works 100% of the time. Purely inspired by your project I'm playing with sending data from dialogflow to my script, but I've never used dialogflow before and I certainly don't know how to do all the data massaging that you do. Nevertheless I'll be testing out your tweaks later tonight too.

**Comment by maykar** [comment]
My DialogFlow implementation is probably the simplest version you could do. It's only taking what you say and sending it to HA. All the work is done in the integration. I've been meaning to look into DialogFlow more. For my purposes and the fact that IFTTT is so much simpler, I haven't really messed around with it too much more than that, but I really should.

An interesting feature that should be possible with DialogFlow is room awareness so you wouldn't have to tell Plex Assistant what device to play on. The only issue is the fact that everyone's webhook will be different. I'm don't think you can publish an actual Google Assistant action with a variable webhook and when publishing you wouldn't be able to use the name "Plex" since it's recognized as a company.

**Comment by maykar** [comment]
Well this took an interesting turn. Plex has updated their apps for many devices including Roku. This update managed to break playback of playqueues. So more investigation is on hold until that's fixed.

Let me know if you get the lovely "an error occurred loading items into the play queue" message as well.

**Comment by dnikles** [comment]
Well I think I have something that works, for me, just using automations & scripts in home assistant. I moved back to IFTTT because dialogflow was a bit annoying with the test speech etc. I get a "command" from IFTTT and in a script pull out the movie and the tv
```
variables:
  plex_movie: '{{ command.split(" on ")[0].split("play ")[1] }}'
  plex_player: '{{ command.split(" on ")[1] }}'
```

So plex_movie might be elf and plex_player might be "kitchen tv". Really rudimentary parsing. If I have a movie with the word "on" in it I'll be out of luck.  But I need to change kitchen tv to media_player.plex_plex_for_roku_kitchen_tv and also to media_player.kitchen_tv so I manually map it in this script (choose action and for each option, maps a value to  plex_entity and player_entity) I then pass plex_entity, player_entity, and movie to another script. 

```
alias: plex_player
sequence:
  - choose:
      - conditions:
          - condition: template
            value_template: '{{ states(player_entity) == ''standby'' }}'
        sequence:
          - service: media_player.turn_on
            data:
              entity_id: '{{ player_entity }}'
    default: []
  - choose:
      - conditions:
          - condition: template
            value_template: >-
              {{ state_attr(player_entity,'source') == 'Plex - Stream for Free'
              }}
        sequence: []
    default:
      - service: media_player.select_source
        data:
          entity_id: '{{ player_entity }}'
          source: Plex - Stream for Free
  - repeat:
      while:
        - condition: template
          value_template: '{{ states(plex_entity) == ''unavailable'' }}'
      sequence:
        - service: plex.scan_for_clients
          data: {}
        - delay:
            hours: 0
            minutes: 0
            seconds: 1
            milliseconds: 0
  - service: media_player.play_media
    data:
      entity_id: '{{ plex_entity }}'
      media_content_type: movie
      media_content_id: '{ "library_name": "Movies", "title": "{{ movie }}" }'
mode: single
```
This script checks if a player is in standby, if it is then turns it on. Then it checks if plex is running, if it isn't it launches plex. Then it checks if the plex client entity is available in home assistant. If it's unavailable, it will scan for clients, wait a second,  then check the status again. After that it will play the movie on the plex client. It's not nearly as modular as your solution, but it's working for me. I never would've gotten to this point without your help.

**Comment by maykar** [comment]
Very cool and that gives me a couple of ideas. 
* I could make start scripts wait for the script to finish before continuing, allowing for the script itself to do a "while unavailable, scan for clients"
* Your script could call the plex_assistant.command instead of play_media

**Comment by maykar** [comment]
Updated both [`__init__.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/__init__.py) and [`helpers.py`](https://raw.githubusercontent.com/maykar/plex_assistant/dev/custom_components/plex_assistant/helpers.py) as well as the [start script example](https://github.com/maykar/plex_assistant/blob/dev/README.md#start-scripts).

Used an early example script from a comment of yours as a template for the example. No longer use a responding check and put the heavy lifting in the script instead.

If you haven't completely given up on it, I'd appreciate your thoughts and/or a test.

**Comment by maykar** [comment]
In my testing this works well in all cases, even when a device is falsely listed as available (as long as you give it a few seconds after the app closes).

**Comment by dnikles** [comment]
going to test it out some more, had to move some stuff around to make sure my method wasn't stomping on yours. I think it worked but I'll have to test some more. I think I ran into a possible issue with my start script. I think if I turn off my tv while it is on plex, the source will stay plex even though the tv is off. The script checks that the tv is off, then turns it on, but next thinks plex is running even if it isn't. Easy enough fix, just add the action to switch to plex after the tv turns on without checking it. If the tv was already on, we can still do the plex check and change source if needed. 

Well I tested out some more before I got to submit this. I had to go back and change my start script to do the while loop to wait to see that the plex integration saw the client before I stopped doing the check for clients. Then in the logs, it looked like everything worked, but the movie did not play for me. I might look at the work plex assistant does in parsing the movie and player, finding the movie, but then try to call my playback script, because for whatever reason that's what seems to work for me.

**Comment by maykar** [comment]
Maybe using the example script with a delay after the device is "found".

Ignoring things like: checking against roman numerals and numbers as ordinals, stripping the string of filler words, checking against words/phrases in localize.py, and separating out sections of the command as needed, etc. When Plex Assistant processes the command it uses a python library called FuzzyWuzzy that checks a string against an array of strings (movie titles, show titles, device names, strings in localize.py, etc). FuzzyWuzzy then returns the best match and a score for the match.

So there is a fair bit of comparing the scores given to decide if what you're saying is a media title, command, or device, as well as picking out keywords to decide which library (Movies, TV, and soon Music) to then search in. It then does similar processing with the filtering keywords `season, episode, latest, recent, new, unwatched, next, ondeck, random, etc` to further refine your search to the item you're looking for.

**Comment by maykar** [comment]
[process_speech.py](https://github.com/maykar/plex_assistant/blob/master/custom_components/plex_assistant/process_speech.py) and the [find_media function](https://github.com/maykar/plex_assistant/blob/4574300a3a90fe1b8758e4f9f50411df2a0a5a29/custom_components/plex_assistant/helpers.py#L296) is where the majority of this happens.

**Comment by maykar** [comment]
The new example script in the readme uses the loop to wait for the device to become available and Plex Assistant now waits until the script is finished before continuing.

**Comment by dnikles** [comment]
I just got around to testing this again. The first few times I tried it, I got "playback error: there are no playable items" on my client, but no errors in the logs. Then I tried calling it with a different movie and it worked, then I tried calling with the original movie and it worked that time. 

I noticed a delay of about 30 seconds between where it lists the playback device in the log and where it lists the movie 
```
2021-02-27 09:19:31 DEBUG (SyncWorker_3) [custom_components.plex_assistant.const] Device: Bedroom TV
2021-02-27 09:20:03 DEBUG (SyncWorker_3) [custom_components.plex_assistant.const] Media: <Movie:4499:Home-Alone>
```
Is that delay built in plex assistant somewhere, in waiting for my start script to complete, or maybe in slowness of my plex server in searching for the movie? My poor little plex server is currently dvring some programs and being played back on another tv in the house as I'm testing this. It's over ten years old on an atom powered nas and I'm surprised it even works at all at this point, but I'm actually replacing it this weekend. I'll test things out later this week once the new server is in place. I wouldn't be surprised if both issues I saw today are related to my overtaxed plex server.

**Comment by maykar** [comment]
There is no delay in Plex Assistant with start scripts anymore, it just waits for your script to finish and then continues. So, any delay would be in the script itself, depending on the script that can be the time it takes for the device to report as available or for any built in delays (like the one commented out at the end of the example).

**Comment by maykar** [comment]
Oh, sorry. Misunderstood. I see where the delay is happening there. And you're correct in assuming that it would be coming from response from the Plex server. I also have an update coming out soon that will include optimizations for searches and other things that should help with speed as well, but a 30 second delay is quite long and I would assume that is a due to the speed of the server.

**Comment by dnikles** [comment]
That's great news for me, thanks for the explanation!

**Comment by dnikles** [comment]
I got everything migrated over to a new plex server. Everything is working, but I still see a bit more of a delay than I was expecting
```
2021-02-28 11:09:36 DEBUG (SyncWorker_5) [custom_components.plex_assistant.const] IFTTT Call: play Captain Underpants on bedroom TV
2021-02-28 11:09:36 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=plex_assistant, service=command, service_data=command=play Captain Underpants on bedroom TV>
2021-02-28 11:09:36 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] Command: play Captain Underpants on bedroom TV
2021-02-28 11:09:44 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] Processed Command: {'media': 'captain underpants', 'device': 'bedroom tv'}
2021-02-28 11:09:52 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] PA Devices: {it's in here}
2021-02-28 11:09:52 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] Device: Bedroom Tv
2021-02-28 11:10:32 DEBUG (SyncWorker_2) [custom_components.plex_assistant.const] Media: <Movie:6222:Captain-Underpants:->
```
I added a notification in my start script, and it is comleting just a moment after I see Device: Bedroom Tv listed in the log, but it was still 40 more seconds until the movie started playing. I don't think the lag is on my plex server anymore, so maybe your upcoming optimization on searches will help. It's a minor issue, but when I was using the playback script [here](https://github.com/maykar/plex_assistant/issues/93#issuecomment-781756644) it would start playing a movie within a second or two of plex launching.

**Comment by maykar** [comment]
The difference between Plex Assistant and your script is that your script is telling Plex exactly what to play immediately with no processing needed, but if Google misunderstands you it will fail and it is limited to only playing something via title.

Plex Assistant is doing fuzzy searches against what's in your library and what you're saying so that it can do the more advanced things like [filtering of items](https://github.com/maykar/plex_assistant#filter-keywords), continuous playing of items returned, finding partial matches, correcting something google has misheard, etc.

I released the latest version a little while ago (1.1.4), would be interested to hear if it helps any. Do you have a fairly large Plex library?

**Comment by dnikles** [comment]
maybe the slowdown is on my home assistant install then? I just have it on a raz pi 3b. I just moved over to the hacs install so I can get notification of new releases and I'm seeing about the same performance. I understand your method does a lot more legwork than mine was, which I much prefer and will stick with what you're doing. I don't have a particularly large movie library, less than 200 movies. Maybe I'll move my HA install over to docker on the new host.

**Comment by maykar** [comment]
Would be interested if the latest version has had any noticeable effect. I've changed the fuzzy searching library to one that performs much better and am utilizing caching on library items now as well. I fear that the 1gb of ram of the 3b might be the limiting factor here, but if possible I'd like to make even low end systems viable.

**Comment by dnikles** [comment]
I'll test it out. I found that something was pegging my processor and a HA core restart seems to have resolved it and it hasn't come back, so I think the problem may have been more localized to my server at the time than anything to do with plex assistant.

**Comment by maykar** [comment]
Good to know, but at least it caused me to make this even faster 😄
